File: backends/elixir-phoenix/lib/stackverse_backend/i18n.ex
defmodule StackverseBackend.I18n do
  @moduledoc false

  import Ecto.Query

  alias StackverseBackend.Repo
  alias StackverseBackend.Schemas.Message

  @default_language "en"

  def default_language, do: @default_language

  def resolve_language(lang, accept_language) do
    supported = supported_languages()

    cond do
      is_binary(lang) and MapSet.member?(supported, lang) ->
        lang

      true ->
        accept_language
        |> parse_accept_language()
        |> Enum.find(&MapSet.member?(supported, &1))
        |> Kernel.||(@default_language)
    end
  end

  def parse_accept_language(nil), do: []
  def parse_accept_language(""), do: []

  def parse_accept_language(header) do
    header
    |> String.split(",", trim: true)
    |> Enum.with_index()
    |> Enum.map(fn {part, index} ->
      [tag | parameters] = String.split(String.trim(part), ";")

      quality =
        parameters
        |> Enum.find_value(1.0, fn parameter ->
          case Regex.run(~r/^\s*q=([0-9.]+)\s*$/, parameter) do
            [_, raw] ->
              case Float.parse(raw) do
                {value, ""} -> value
                _ -> 0.0
              end

            _ ->
              nil
          end
        end)

      code =
        tag
        |> String.downcase()
        |> String.split("-", parts: 2)
        |> hd()

      {code, quality, index}
    end)
    |> Enum.filter(fn {code, _quality, _index} -> Regex.match?(~r/^[a-z]{1,8}$/, code) end)
    |> Enum.sort_by(fn {_code, quality, index} -> {-quality, index} end)
    |> Enum.map(fn {code, _quality, _index} -> code end)
  end

  def localize(key, language) do
    text_for(key, language) || text_for(key, @default_language) || key
  end

  def bundle(language) do
    languages = Enum.uniq([language, @default_language])

    messages = Message |> where([message], message.language in ^languages) |> Repo.all()

    fallback =
      messages
      |> Enum.filter(&(&1.language == @default_language))
      |> Map.new(&{&1.key, &1.text})

    requested =
      messages
      |> Enum.filter(&(&1.language == language))
      |> Map.new(&{&1.key, &1.text})

    Map.merge(fallback, requested)
  end

  defp supported_languages do
    Message
    |> distinct([message], message.language)
    |> select([message], message.language)
    |> Repo.all()
    |> MapSet.new()
  end

  defp text_for(key, language) do
    Repo.one(
      from message in Message,
        where: message.key == ^key and message.language == ^language,
        select: message.text
    )
  end
end

File: backends/scala-http4s/src/main/scala/dev/stackverse/http4s/I18n.scala
package dev.stackverse.http4s

import io.circe.{Json, JsonObject}

import scala.util.Try

final class I18n(db: Db) extends ProblemLocalization {
  val DefaultLanguage = "en"

  override def resolve(queryLang: Option[String], acceptLanguage: Option[String]): String =
    db.withConnection { conn =>
      val supported = db.query(conn, "select distinct language from messages")(rs => rs.getString("language")).toSet
      queryLang
        .filter(supported.contains)
        .orElse(parseAcceptLanguage(acceptLanguage).find(supported.contains))
        .getOrElse(DefaultLanguage)
    }

  override def localize(key: String, language: String): String =
    db.withConnection { conn =>
      db.one(
        conn,
        "select text from messages where key = ? and language = any(?::text[]) order by case when language = ? then 0 else 1 end limit 1",
        Seq(key, Seq(language, DefaultLanguage).distinct, language)
      )(_.getString("text"))
        .getOrElse(key)
    }

  def bundle(language: String): JsonObject =
    db.withConnection { conn =>
      val rows = db.query(
        conn,
        "select key, language, text from messages where language = any(?::text[]) order by key",
        Seq(Seq(language, DefaultLanguage).distinct)
      ) { rs =>
        (rs.getString("key"), rs.getString("language"), rs.getString("text"))
      }
      val values = scala.collection.mutable.LinkedHashMap.empty[String, String]
      rows.foreach { case (key, lang, text) =>
        if (lang == language || !values.contains(key)) values.update(key, text)
      }
      JsonObject.fromIterable(values.toSeq.map { case (key, text) => key -> Json.fromString(text) })
    }

  private def parseAcceptLanguage(header: Option[String]): Seq[String] =
    header.toSeq
      .flatMap(_.split(",").toSeq.zipWithIndex)
      .flatMap { case (part, index) =>
        val pieces = part.trim.split(";").map(_.trim)
        val code = pieces.headOption.getOrElse("").toLowerCase.split("-").headOption.getOrElse("")
        val q = pieces
          .drop(1)
          .collectFirst {
            case value if value.startsWith("q=") => Try(value.stripPrefix("q=").toDouble).getOrElse(0.0)
          }
          .getOrElse(1.0)
        if (code.matches("^[a-z]{1,8}$")) Some((code, q, index)) else None
      }
      .sortBy { case (_, q, index) => (-q, index) }
      .map(_._1)
}
